### PR TITLE
Simplify service invocation APIs

### PR DESCRIPTION
**Pull request by rynowak**
# Description

This change mixes up the service invocation APIs to reflect some changes
that are coming down the pipe in dapr/dapr#2683.

We're now using HTTP's semantics and JSON for the main functionality in
`InvokeMethodAsync`. This means that the error handling is simplified
and base on HTTP's status codes rather than trying to abstract over HTTP
and gRPC's different error models. This also means that we're free to
expose HttpRequestMessage and related types in the API because it's no
longer an abstraction.

You can still use `InvokeMethodAsync` to invoke gRPC services, but you
will get a status code in the response, and it will expect JSON, etc,
similar to how gRPC gateway works.

In addition to that we're added `InvokeMethodGrpcAsync` for cases where
you really want *gRPC's* semantics. These methods do Protobuf
serialization, as is a natural choice for gRPC. I do not recommend using
this approach to invoke HTTP services.

---

I also make updates and improvments to various infrastructure where
necessary (like InvocationException). Also, the gRPC sample in the repo
now uses the new `InvokeMethodGrpcAsync` APIs.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Fixes: #549

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

### CODE REVIEW

File: src/Dapr.Client/DaprClientBuilder.cs [hunk 1]
```diff
@@ -16,84 +17,119 @@ namespace Dapr.Client
...
+        public DaprClientBuilder UseHttpEndpoint(string daprEndpoint)
         {
-            var daprGrpcPort = Environment.GetEnvironmentVariable("DAPR_GRPC_PORT") ?? DefaultDaprGrpcPort;
-            this.daprEndpoint = $"http://127.0.0.1:{daprGrpcPort}";
-            this.jsonSerializerOptions = new JsonSerializerOptions
-            {
-                PropertyNameCaseInsensitive = true
-            };
+            ArgumentVerifier.ThrowIfNullOrEmpty(daprEndpoint, nameof(daprEndpoint));
+            this.DaprGrpcEndpoint = daprEndpoint;
```
Comment (vinayada1) [inline src/Dapr.Client/DaprClientBuilder.cs:67]: this. DaprHttpEndpoint instead of grpc

File: src/Dapr.Client/DaprClientGrpc.cs [hunk 2]
```diff
@@ -210,124 +197,177 @@ private static T ConvertFromInvokeBindingResponse<T>(Autogenerated.InvokeBinding
...
-            CallOptions callOptions;
-            (request, callOptions) = this.MakeInvokeRequestAsync(appId, methodName, null, httpOptions, cancellationToken);
-            var response = await client.InvokeServiceAsync(request, callOptions);
-            return response.Data.Value.IsEmpty ? default : TypeConverters.FromAny<TResponse>(response.Data, this.jsonSerializerOptions);
+            var envelope = new Autogenerated.InvokeServiceRequest()
+            {
+                Id = appId,
+                Message = new Autogenerated.InvokeRequest()
+                {
+                    Method = methodName,
```
Comment (vinayada1) [inline src/Dapr.Client/DaprClientGrpc.cs:267]: nit: extra whitespace

File: src/Dapr.Client/DaprClientBuilder.cs (hunk 1 above)
Comment (vinayada1) [reply src/Dapr.Client/DaprClientBuilder.cs:67]: add UT

File: src/Dapr.Client/DaprClientGrpc.cs [hunk 3]
```diff
@@ -210,124 +197,177 @@ private static T ConvertFromInvokeBindingResponse<T>(Autogenerated.InvokeBinding
...
             ArgumentVerifier.ThrowIfNullOrEmpty(appId, nameof(appId));
-            ArgumentVerifier.ThrowIfNullOrEmpty(methodName, nameof(methodName));
+            ArgumentVerifier.ThrowIfNull(methodName, nameof(methodName));
 
-            var (request, callOptions) = this.MakeInvokeRequestAsync(appId, methodName, null, httpOptions, cancellationToken);
-            await client.InvokeServiceAsync(request, callOptions);
+            // Note about this, it's possible to construct invalid stuff using path navigation operators
+            // like `../..`. But the principle of garbage in -> garbage out holds.
+            //
+            // This approach leads avoids some common pitfals that could lead to undesired encoding.
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Dapr.Client/DaprClientGrpc.cs:210]: One pitfall of writing comments is misspelling words. This leads to extra code review comments :)

File: src/Dapr.Client/DaprClientGrpc.cs [hunk 4]
```diff
@@ -210,124 +197,177 @@ private static T ConvertFromInvokeBindingResponse<T>(Autogenerated.InvokeBinding
...
-           string methodName,
-           HttpInvocationOptions httpOptions = default,
-           CancellationToken cancellationToken = default)
+
+        public override HttpRequestMessage CreateInvokeMethodRequest(HttpMethod httpMethod, string appId, string methodName)
         {
+            ArgumentVerifier.ThrowIfNull(httpMethod, nameof(httpMethod));
             ArgumentVerifier.ThrowIfNullOrEmpty(appId, nameof(appId));
-            ArgumentVerifier.ThrowIfNullOrEmpty(methodName, nameof(methodName));
+            ArgumentVerifier.ThrowIfNull(methodName, nameof(methodName));
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Dapr.Client/DaprClientGrpc.cs:205]: For HTTP semantics, the empty string would map to the root URL. This is allowed by the runtime and is a somewhat important scenario. 

For gRPC semantics this should require non-empty. I'll double-check these.

File: src/Dapr.Client/DaprClient.cs [hunk 5]
```diff
@@ -163,101 +170,344 @@ public static HttpClient CreateInvokeHttpClient(string appId = null, string dapr
...
+        /// If the response has a non-success status code an exception will be thrown.
         /// </summary>
         /// <param name="appId">The Dapr application id to invoke the method on.</param>
-        /// <param name="methodName">The name of the method to invoke.</param>  
-        /// <param name="data">Byte array to pass to the method</param>      
-        /// <param name="httpOptions">Additional fields that may be needed if the receiving app is listening on HTTP.</param>
+        /// <param name="methodName">The name of the method to invoke.</param>
         /// <param name="cancellationToken">A <see cref="CancellationToken" /> that can be used to cancel the operation.</param>
-        /// <returns>A <see cref="Task{T}" /> that will return the value when the operation has completed.</returns>  
-        public abstract Task<InvocationResponse<byte[]>> InvokeMethodRawAsync(
```
Comment (vinayada1) [inline src/Dapr.Client/DaprClient.cs:254]: Don't we need the raw anymore?

Comment (rynowak) [reply src/Dapr.Client/DaprClient.cs:254]: We don't need a special method for it anymore - you can use `CreateInvokeMethodRequest()` and then do whatever you like.

File: src/Dapr.Client/DaprClientBuilder.cs (hunk 1 above)
Comment (rynowak) [reply src/Dapr.Client/DaprClientBuilder.cs:67]: Adding validation in build where we convert to URIs

Comment (rynowak) [reply src/Dapr.Client/DaprClientBuilder.cs:67]: > add UT 

For a method that sets a property? Not worth the 3 minutes it would take.

Comment (rynowak) [reply src/Dapr.Client/DaprClientBuilder.cs:67]: I'll add tests for the new validation.

File: src/Dapr.Client/DaprClientBuilder.cs [hunk 6]
```diff
@@ -16,84 +17,119 @@ namespace Dapr.Client
...
+        /// The URI endpoint to use for gRPC calls to the Dapr runtime. The default value will be 
+        /// <c>http://127.0.0.1:DAPR_GRPC_PORT</c> where <c>DAPR_GRPC_PORT</c> represents the value of the 
+        /// <c>DAPR_GRPC_PORT</c> environment variable.
+        /// </param>
+        /// <returns>The <see cref="DaprClientBuilder" /> instance.</returns>
+        public DaprClientBuilder UseGrpcEndpoint(string daprEndpoint)
         {
             ArgumentVerifier.ThrowIfNullOrEmpty(daprEndpoint, nameof(daprEndpoint));
-            this.daprEndpoint = daprEndpoint;
+            this.DaprGrpcEndpoint = daprEndpoint;
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Dapr.Client/DaprClientBuilder.cs:83]: Adding unit tests for the new validation. We shouldn't bother adding unit tests for trivial functionality like this, that time is better spend on almost anything else.

File: src/Dapr.Client/DaprClientGrpc.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Dapr.Client/DaprClientGrpc.cs:205]: I double-checked these and they are correct. The HTTP ones allow an "empty" method name, the gRPC ones don't.

File: src/Dapr.Client/DaprClientGrpc.cs [hunk 7]
```diff
@@ -210,124 +197,177 @@ private static T ConvertFromInvokeBindingResponse<T>(Autogenerated.InvokeBinding
...
-           string methodName,
-           TRequest data,
-           HttpInvocationOptions httpOptions = default,
-           CancellationToken cancellationToken = default)
+        public override HttpRequestMessage CreateInvokeMethodRequest<TRequest>(HttpMethod httpMethod, string appId, string methodName, TRequest data)
         {
+            ArgumentVerifier.ThrowIfNull(httpMethod, nameof(httpMethod));
             ArgumentVerifier.ThrowIfNullOrEmpty(appId, nameof(appId));
-            ArgumentVerifier.ThrowIfNullOrEmpty(methodName, nameof(methodName));
+            ArgumentVerifier.ThrowIfNull(methodName, nameof(methodName));
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Dapr.Client/DaprClientGrpc.cs:219]: Answered on the other comment.

File: src/Dapr.Client/DaprClientGrpc.cs (hunk 2 above)
Comment (rynowak) [reply src/Dapr.Client/DaprClientGrpc.cs:267]: I don't see any, I autoformatted the document and it's still in the same place.

**Comment by rynowak** [comment]
Updated

File: src/Dapr.Client/InvocationException.cs [hunk 8]
```diff
@@ -2,27 +2,58 @@
...
+
+        /// <summary>
+        /// Gets the <see cref="HttpResponseMessage" /> of the request that failed. Will be <c>null</c> if the 
+        /// failure was not related to an HTTP request or preventing the response from being recieved.
+        /// </summary>
+        public HttpResponseMessage Response { get; }
+
+        private static string FormatExceptionForFailedRequest(string appId, string methodName)
+        {
+            return $"An exception occurred while invoking method: '{methodName}' on app-d: '{appId}'";
```
Comment (vinayada1) [inline src/Dapr.Client/InvocationException.cs:56]: typo: app-id

Comment (rynowak) [reply src/Dapr.Client/InvocationException.cs:56]: OH NO

File: src/Dapr.Client/DaprClientGrpc.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (vinayada1) [reply src/Dapr.Client/DaprClientGrpc.cs:210]: A pitfall of replying to comments is that it is still pitfal :)

**Review by vinayada1 — APPROVED** [review]

**Pull request merged**